File: Template method/c#/TemplateMethod.cs
abstract class Transmitter
{
    protected virtual void VoiceRecord() => Console.WriteLine("recording a fragment of speech");

    protected virtual void Simpling() {}
    
    protected virtual void Digitization() {}
    
    protected virtual void Modulation() {}
    
    protected virtual void Transmission() => Console.WriteLine("signal transmission via radio channel");

    public void ProcessStart()
    {
        VoiceRecord();
        Simpling();
        Digitization();
        Modulation();
        Transmission();
    }
}

class AnalogTransmitter : Transmitter
{
    protected override void Modulation() => Console.WriteLine("modulation of the analog signal");
}

class DigitTransmitter : Transmitter
{
    protected override void Simpling() => Console.WriteLine("sampling of the recorded fragment");
    
    protected override void Digitization() => Console.WriteLine("digitization");
    
    protected override void Modulation() => Console.WriteLine("digital signal modulation");
}

class Program
{
    static void Main()
    {
        Transmitter analogTransmitter = new AnalogTransmitter();
        analogTransmitter.ProcessStart();
        
        Console.WriteLine();

        Transmitter digitTransmitter = new DigitTransmitter();
        digitTransmitter.ProcessStart();
    }
}

File: Template method/java/TemplateMethod.java
abstract class Transmitter{
    protected void voiceRecord(){
        System.out.println("recording a fragment of speech");
    }

    protected void simpling(){

    }

    protected void digitization(){

    }

    protected void modulation(){

    }

    protected void transmission(){
        System.out.println("signal transmission via radio channel");
    }

    public void ProcessStart(){
        voiceRecord();
        simpling();
        digitization();
        modulation();
        transmission();
    }
}

class AnalogTransmitter extends Transmitter{
    @Override
    protected void modulation(){
        System.out.println("modulation of the analog signal");
    }
}

class DigitTransmitter extends Transmitter{
    @Override
    protected void simpling(){
        System.out.println("sampling of the recorded fragment");
    }

    @Override
    protected void digitization(){
        System.out.println("digitization");
    }
    
    @Override
    protected void modulation(){
        System.out.println("digital signal modulation");
    }
}

public class Main {
    public static void main(){
        
        Transmitter analogTransmitter = new AnalogTransmitter();
        analogTransmitter.ProcessStart();

        System.out.println();

        Transmitter digitTransmitter = new DigitTransmitter();
        digitTransmitter.ProcessStart();
    }
}